src/eval.c

```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    {
+	if ((p[-1] == '$' && p[0] == '$') || (p[0] == '$' && p[1] == '$'))  // '$$' escapes to '$'
+	{
+	    continue;
+	}
+	else if (p[0] == '$')
+	{
+	    int embedded_is_literal = (p[1] == '{');
+	    int embedding_start_length;
+
+	    embedded[finished] = embedded_is_literal
+		? read_next_embedded_lit(p, template)
+		: read_next_embedded_var(p, template);
+	    if (embedded[finished] == NULL)
+	    {
+		free(embedded);
+		return NULL;
+	    }
+	    embedding_start_length =
+		(embedded_is_literal ? LITERAL_EMBEDDING_START_LENGTH : VAR_EMBEDDING_START_LENGTH)
+		+ STRLEN(embedded[finished]->expr);
+
+	    mb_ptr_adv_nth(&p, embedding_start_length);  // the part of "${literal" or "$var"
+	    ++finished;
+	    continue; // the part of "}" (increasing)
+	}
+    }
+
+    return embedded;
+}
+
+    static int
+read_next_string_length(char_u *whole_string, int expect_literal_string, char_u *template)
+{
+    char_u*	p;
+    int		(*is_closing_quote)(char_u*);
+
+    // If entered with the closing quote of a template string into this function
+    if (whole_string[1] == NUL)
+    {
+	semsg(_("E1000: Unterminated template literal: %s"), template);
+	return -1;
+    }
+
+    p = whole_string + 1;  // Skip the head of '"' (or '\'')
+    is_closing_quote = expect_literal_string
+	? is_closing_single_quote
+	: is_closing_double_quote;
+
+    for (; !is_closing_quote(p); MB_PTR_ADV(p))
+    {
+	if (*p == NUL)
+	{
+	    semsg(_("E115: Missing quote: %s"), template);
+	    return -1;
+	}
+    }
+
+    return p - whole_string;
+}
+
+/*
+ * Get a first "expr" of "${" "expr" "}" from `whole_embedding`.
+ *
+ * whole_embedding starts with "${": e.g. "${10}`", "${var} Hello :D`".
+ *
+ * The result MUST BE `free`d later! (if it is not NULL.)
+ */
+    static embedded_t*
+read_next_embedded_lit(char_u *whole_embedding, char_u *template)
+{
+    const size_t	LITERAL_EMBEDDING_START_LENGTH = 2;  // "${"
+    char_u*		embedded_start = whole_embedding + LITERAL_EMBEDDING_START_LENGTH;
+    size_t		expr_length = 0;  // "${" expr_length "}"
+    size_t		nested_blocks = 0;  // {}'s nests: ${{'x': 10}}, ${something({})}
+    char_u*		p;
+
+    for (p = embedded_start; !(nested_blocks == 0 && *p == '}'); MB_PTR_ADV(p))
+    {
+	switch (*p)
+	{
+	    case '{':
+		++nested_blocks;
+		break;
+	    case '}':
+		--nested_blocks;
+		break;
+	    case '"':
+	    case '\'':
+	    {
+		int length = read_next_string_length(p, (*p == '\''), template);
+		if (length == -1)
+		{
+		    return NULL;
+		}
+		expr_length += length;
+		mb_ptr_adv_nth(&p, length);
+		break;
+	    }
+	    case NUL:
+		semsg(_("E1000: Unterminated template literal: %s"), template);
+		return NULL;  // failed! a closing "}" couldn't be found.
+	}
+	++expr_length;
+    }
+
+    return new_embedded(TRUE, embedded_start, expr_length);
+}
+
+/*
+ * Get a first "var" of "$" "var" from `whole_embedding`.
+ *
+ * whole_embedding starts with "$": e.g. "$var`", "$v_i_m".
```

Comment:
The backquote can be removed?

Reply:
:+1:

Reply:
I meant that ```"$var`"``` isn't a mistake of ```"$var"```?

Reply:
Sorry, my reply was too meaningless x(
`` "$var` `` is not intended, but `"$var'"` is intended.

I replaced to
```
 * whole_embedding starts with "$": e.g. "$var", "$var'", "$v_i_m".
```
:+1: